File: applications/rtkdualenergyforwardmodel/rtkdualenergyforwardmodel.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkdualenergyforwardmodel_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"
#include "rtkMacro.h"
#include "rtkSpectralForwardModelImageFilter.h"
#include "rtkConstantImageSource.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkdualenergyforwardmodel, args_info);

  using PixelValueType = float;
  constexpr unsigned int Dimension = 3;

  using DecomposedProjectionType = itk::VectorImage<PixelValueType, Dimension>;
  using DecomposedProjectionReaderType = itk::ImageFileReader<DecomposedProjectionType>;

  using DualEnergyProjectionsType = itk::VectorImage<PixelValueType, Dimension>;
  using DualEnergyProjectionWriterType = itk::ImageFileWriter<DualEnergyProjectionsType>;

  using IncidentSpectrumImageType = itk::Image<PixelValueType, Dimension>;
  using IncidentSpectrumReaderType = itk::ImageFileReader<IncidentSpectrumImageType>;

  using DetectorResponseImageType = itk::Image<PixelValueType, Dimension - 1>;
  using DetectorResponseReaderType = itk::ImageFileReader<DetectorResponseImageType>;

  using MaterialAttenuationsImageType = itk::Image<PixelValueType, Dimension - 1>;
  using MaterialAttenuationsReaderType = itk::ImageFileReader<MaterialAttenuationsImageType>;

  // Read all inputs
  auto decomposedProjectionReader = DecomposedProjectionReaderType::New();
  decomposedProjectionReader->SetFileName(args_info.input_arg);
  decomposedProjectionReader->Update();

  auto incidentSpectrumReaderHighEnergy = IncidentSpectrumReaderType::New();
  incidentSpectrumReaderHighEnergy->SetFileName(args_info.high_arg);
  incidentSpectrumReaderHighEnergy->Update();

  auto incidentSpectrumReaderLowEnergy = IncidentSpectrumReaderType::New();
  incidentSpectrumReaderLowEnergy->SetFileName(args_info.low_arg);
  incidentSpectrumReaderLowEnergy->Update();

  auto materialAttenuationsReader = MaterialAttenuationsReaderType::New();
  materialAttenuationsReader->SetFileName(args_info.attenuations_arg);
  materialAttenuationsReader->Update();

  // Get parameters from the images
  const unsigned int MaximumEnergy =
    incidentSpectrumReaderHighEnergy->GetOutput()->GetLargestPossibleRegion().GetSize(0);

  // If the detector response is given by the user, use it. Otherwise, assume it is included in the
  // incident spectrum, and fill the response with ones
  auto                               detectorResponseReader = DetectorResponseReaderType::New();
  DetectorResponseImageType::Pointer detectorImage;
  if (args_info.detector_given)
  {
    detectorResponseReader->SetFileName(args_info.detector_arg);
    detectorResponseReader->Update();
    detectorImage = detectorResponseReader->GetOutput();
  }
  else
  {
    rtk::ConstantImageSource<DetectorResponseImageType>::Pointer detectorSource =
      rtk::ConstantImageSource<DetectorResponseImageType>::New();
    detectorSource->SetSize(itk::MakeSize(1, MaximumEnergy));
    detectorSource->SetConstant(1.0);
    detectorSource->Update();
    detectorImage = detectorSource->GetOutput();
  }

  // Generate a set of zero-filled intensity projections
  auto dualEnergyProjections = DualEnergyProjectionsType::New();
  dualEnergyProjections->CopyInformation(decomposedProjectionReader->GetOutput());
  dualEnergyProjections->SetVectorLength(2);
  dualEnergyProjections->Allocate();

  // Check that the inputs have the expected size
  DecomposedProjectionType::IndexType indexDecomp;
  indexDecomp.Fill(0);
  if (decomposedProjectionReader->GetOutput()->GetVectorLength() != 2)
    itkGenericExceptionMacro(<< "Decomposed projections (i.e. initialization data) image has vector length "
                             << decomposedProjectionReader->GetOutput()->GetVectorLength() << ", should be 2");

  if (materialAttenuationsReader->GetOutput()->GetLargestPossibleRegion().GetSize()[1] != MaximumEnergy)
    itkGenericExceptionMacro(<< "Material attenuations image has "
                             << materialAttenuationsReader->GetOutput()->GetLargestPossibleRegion().GetSize()[1]
                             << "energies, should have " << MaximumEnergy);

  // Create and set the filter
  using ForwardModelFilterType =
    rtk::SpectralForwardModelImageFilter<DecomposedProjectionType, DualEnergyProjectionsType>;
  auto forward = ForwardModelFilterType::New();
  forward->SetInputDecomposedProjections(decomposedProjectionReader->GetOutput());
  forward->SetInputMeasuredProjections(dualEnergyProjections);
  forward->SetInputIncidentSpectrum(incidentSpectrumReaderHighEnergy->GetOutput());
  forward->SetInputSecondIncidentSpectrum(incidentSpectrumReaderLowEnergy->GetOutput());
  forward->SetDetectorResponse(detectorImage);
  forward->SetMaterialAttenuations(materialAttenuationsReader->GetOutput());
  forward->SetIsSpectralCT(false);
  forward->SetComputeVariances(args_info.variances_given);

  TRY_AND_EXIT_ON_ITK_EXCEPTION(forward->Update())

  // Write output
  auto writer = DualEnergyProjectionWriterType::New();
  writer->SetInput(forward->GetOutput());
  writer->SetFileName(args_info.output_arg);
  writer->Update();

  // If requested, write the variances
  if (args_info.variances_given)
  {
    TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(forward->GetOutputVariances(), args_info.variances_arg))
  }

  return EXIT_SUCCESS;
}

File: applications/rtkdualenergyforwardmodel/rtkdualenergyforwardmodel.py
#!/usr/bin/env python
import argparse
import sys
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Computes expected photon counts from incident spectrum, material attenuations, detector response and material-decomposed projections"
    )

    # General options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--output", "-o", help="Output file name (high and low energy projections)", type=str, required=True
    )
    parser.add_argument(
        "--input", "-i", help="Material-decomposed projections", type=str, required=True
    )
    parser.add_argument(
        "--high", help="Incident spectrum image, high energy", type=str, required=True
    )
    parser.add_argument(
        "--low", help="Incident spectrum image, low energy", type=str, required=True
    )
    parser.add_argument(
        "--detector", "-d", help="Detector response file", type=str
    )
    parser.add_argument(
        "--attenuations", "-a", help="Material attenuations file", type=str, required=True
    )
    parser.add_argument(
        "--variances", help="Output variances of measured energies, file name", type=str
    )

    return parser


def process(args_info: argparse.Namespace):
    # Define pixel type and dimension
    PixelValueType = itk.F
    Dimension = 3

    DecomposedProjectionType = itk.VectorImage[PixelValueType, Dimension]
    DualEnergyProjectionsType = itk.VectorImage[PixelValueType, Dimension]
    IncidentSpectrumImageType = itk.Image[PixelValueType, Dimension]
    DetectorResponseImageType = itk.Image[PixelValueType, Dimension - 1]
    MaterialAttenuationsImageType = itk.Image[PixelValueType, Dimension - 1]

    # Read all inputs
    if args_info.verbose:
        print("Reading material-decomposed projections...")
    decomposed_projection_reader = itk.ImageFileReader[DecomposedProjectionType].New()
    decomposed_projection_reader.SetFileName(args_info.input)
    decomposed_projection_reader.Update()

    if args_info.verbose:
        print("Reading incident spectrum (high energy)...")
    incident_spectrum_reader_high_energy = itk.ImageFileReader[IncidentSpectrumImageType].New()
    incident_spectrum_reader_high_energy.SetFileName(args_info.high)
    incident_spectrum_reader_high_energy.Update()

    if args_info.verbose:
        print("Reading incident spectrum (low energy)...")
    incident_spectrum_reader_low_energy = itk.ImageFileReader[IncidentSpectrumImageType].New()
    incident_spectrum_reader_low_energy.SetFileName(args_info.low)
    incident_spectrum_reader_low_energy.Update()

    if args_info.verbose:
        print("Reading material attenuations...")
    material_attenuations_reader = itk.ImageFileReader[MaterialAttenuationsImageType].New()
    material_attenuations_reader.SetFileName(args_info.attenuations)
    material_attenuations_reader.Update()

    # Get parameters from the images
    maximum_energy = incident_spectrum_reader_high_energy.GetOutput().GetLargestPossibleRegion().GetSize(0)

    # If the detector response is given by the user, use it. Otherwise, assume it is included in the
    # incident spectrum, and fill the response with ones
    if args_info.detector:
        if args_info.verbose:
            print("Reading detector response...")
        detector_response_reader = itk.ImageFileReader[DetectorResponseImageType].New()
        detector_response_reader.SetFileName(args_info.detector)
        detector_response_reader.Update()
        detector_image = detector_response_reader.GetOutput()
    else:
        if args_info.verbose:
            print("No detector response provided, using default (ones)...")
        detector_source = rtk.ConstantImageSource[DetectorResponseImageType].New()
        detector_source.SetSize([1, maximum_energy])
        detector_source.SetConstant(1.0)
        detector_source.Update()
        detector_image = detector_source.GetOutput()

    # Generate a set of zero-filled intensity projections
    dual_energy_projections = DualEnergyProjectionsType.New()
    dual_energy_projections.CopyInformation(decomposed_projection_reader.GetOutput())
    dual_energy_projections.SetVectorLength(2)
    dual_energy_projections.Allocate()

    # Check that the inputs have the expected size
    if decomposed_projection_reader.GetOutput().GetVectorLength() != 2:
        print(f"Error: Decomposed projections image has vector length {decomposed_projection_reader.GetOutput().GetVectorLength()}, should be 2")
        sys.exit(1)

    if material_attenuations_reader.GetOutput().GetLargestPossibleRegion().GetSize()[1] != maximum_energy:
        print(f"Error: Material attenuations image has {material_attenuations_reader.GetOutput().GetLargestPossibleRegion().GetSize()[1]} energies, should have {maximum_energy}")
        sys.exit(1)

    # Create and set the filter
    if args_info.verbose:
        print("Setting up spectral forward model filter...")
    forward = rtk.SpectralForwardModelImageFilter[DecomposedProjectionType, DualEnergyProjectionsType].New()
    forward.SetInputDecomposedProjections(decomposed_projection_reader.GetOutput())
    forward.SetInputMeasuredProjections(dual_energy_projections)
    forward.SetInputIncidentSpectrum(incident_spectrum_reader_high_energy.GetOutput())
    forward.SetInputSecondIncidentSpectrum(incident_spectrum_reader_low_energy.GetOutput())
    forward.SetDetectorResponse(detector_image)
    forward.SetMaterialAttenuations(material_attenuations_reader.GetOutput())
    forward.SetIsSpectralCT(False)
    forward.SetComputeVariances(args_info.variances is not None)

    forward.Update()

    # Write output
    if args_info.verbose:
        print(f"Writing output to {args_info.output}...")
    writer = itk.ImageFileWriter[DualEnergyProjectionsType].New()
    writer.SetInput(forward.GetOutput())
    writer.SetFileName(args_info.output)
    writer.Update()

    # If requested, write the variances
    if args_info.variances:
        if args_info.verbose:
            print(f"Writing variances to {args_info.variances}...")

        itk.imwrite(forward.GetOutputVariances(), args_info.variances)

    if args_info.verbose:
        print("Processing completed successfully.")


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
